### PR TITLE
Allow running ZooKeeper and KRaft based clusters in parallel

### LINKED ISSUE #9108: Controllers creation can cause metadata lost on ZooKeeper based cluster using node pools
I ran the following scenario ...

Create a ZooKeeper based Kafka cluster. Operator running without node pools nor KRaft feature gates enabled. Create some topics. Going into the shell of any of the brokers I can see the topics (even not using localhost:9092 but the bootstrap service):

```shell
[kafka@my-cluster-kafka-0 kafka]$ bin/kafka-topics.sh --bootstrap-server localhost:9092 --list
my-topic-1
my-topic-2
my-topic-3
```

Now, apply the migration to use the node pools (for running the brokers but still ZooKeeper based) as per the [documentation](https://strimzi.io/docs/operators/latest/deploying#proc-migrating-clusters-node-pools-str).
Everything is fine.

Enable KRaft on the operator. As we know the operator will log this warning which is fine.

```shell
io.strimzi.operator.common.model.InvalidResourceException: Tha Kafka cluster my-cluster is invalid: [At least one KafkaNodePool with the controller role and at least one replica is required when KRaft mode is enabled]
```

Now create a node pool with controllers.
The operator deploys the controllers and, in my test, it was able to roll one of the brokers (configuring it with the connection to the KRaft quorum) but not the other two due to following which would make sense (?).

```shell
2023-09-12 09:14:19 INFO  KafkaRoller:576 - Reconciliation #7(watch) Kafka(kafka/my-cluster): Rolling Pod my-cluster-kafka-0/0 due to [Pod has old revision]                                                                               
2023-09-12 09:14:19 INFO  KafkaAvailability:135 - Reconciliation #7(watch) Kafka(kafka/my-cluster): my-topic-2/0 will be under-replicated (ISR={0,2}, replicas=[0,2,1], min.insync.replicas=2) if broker 0 is restarted.                   
2023-09-12 09:14:19 WARN  KafkaRoller:440 - Reconciliation #7(watch) Kafka(kafka/my-cluster): Pod my-cluster-kafka-0/0 can't be safely force-rolled; original error:                                                                       
2023-09-12 09:14:19 INFO  KafkaRoller:328 - Reconciliation #7(watch) Kafka(kafka/my-cluster): Could not verify pod my-cluster-kafka-0/0 is up-to-date, giving up after 10 attempts. Total delay between attempts 127750ms                  
io.strimzi.operator.cluster.operator.resource.KafkaRoller$ForceableProblem: Pod my-cluster-kafka-0 is controller and there are other pods to verify. Non-controller pods will be verified first 
```

Result ... one broker (the rolled one) is now connected to the KRaft quorum and it has no metadata about previously created topics. Other two are still using ZooKeeper because not rolled. Depending on where your request about metadata lands, you can get that topics don't exist (you hit the rolled broker using KRaft quorum) or they exist (you hit the not-rolled brokers still using ZooKeeper).
For example on a not rolled broker using the bootstrap service ...

```shell
[kafka@my-cluster-kafka-0 kafka]$ bin/kafka-topics.sh --bootstrap-server localhost:9092 --list
my-topic-1
my-topic-2
my-topic-3
[kafka@my-cluster-kafka-0 kafka]$ bin/kafka-topics.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 --list
my-topic-1
my-topic-2
my-topic-3
[kafka@my-cluster-kafka-0 kafka]$ bin/kafka-topics.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 --list
my-topic-1
my-topic-2
my-topic-3
[kafka@my-cluster-kafka-0 kafka]$ bin/kafka-topics.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 --list

[kafka@my-cluster-kafka-0 kafka]
```

Notice last query. It ended to the rolled broker using KRaft with no metadata.
I confirmed it by running the same topics listing command on each broker (rolled and not) by using localhost:9092 as well.

### PR DESCRIPTION
**Pull request by ppatierno**
### Type of change

- Enhancement / new feature

### Description

This PR implements proposal being still discussed here https://github.com/strimzi/proposals/pull/93.
It adds a new annotation `strimzi.io/kraft` to be applied on a `Kafka` custom resource in order to distinguish between ZooKeeper and KRaft based clusters.
It actually changes the behaviour of the `+UseKRaft` feature gate.
With this change, it's not enough enabling the feature gate to create a KRaft based cluster but it's also needed to apply the `strimzi.io/kraft: enabled` annotation.
The `disabled`, missing or any other values identify the cluster as ZooKeeper based and the operator reconciles it in the proper way.

This PR fixes #9108 as well.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Make sure all tests pass
- [x] Update documentation
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Update CHANGELOG.md

### CODE REVIEW

**Review by scholzj — COMMENTED** [review]
You will need to update this in the ST as well I guess?

**Comment by ppatierno** [comment]
@scholzj the build is failing on `KafkaAssemblyOperatorWithPoolsMockTest.testReconcileReplacesAllDeletedSecrets` which works just fine if I run it locally (alone or the entire class) even more times. I cannot find any good reason why it's failing here. Node pools are involved but not KRaft AFAICS. Is it a "known" flaky? Have you already seen this failure?

**Comment by ppatierno** [comment]
@scholzj everything seems to be green :-) Regression was ok, can you have another pass please?

File: CHANGELOG.md
Comment (scholzj) [inline CHANGELOG.md:10]: Maybe worth mentioning also in the `Changes, deprecations and removals` section to make sure users know that they need to change their existing clusters? I would probably:
* Phrase this as `Allow running ZooKeeper and KRaft based Apache Kafka clusters in parallel when the UseKRaft feature gate is enabled`
* List the details about the annotation that needs to be added to the KRaft YAMLs to the changes section

**Review by scholzj — COMMENTED** [review]
One nit to the CHANGELOG changes. But LGTM otherwise. Should it be a Draft until the proposal is approved to make sure we don't merge it by mistake?

File: packaging/install/cluster-operator/060-Deployment-strimzi-cluster-operator.yaml
Comment (scholzj) [inline packaging/install/cluster-operator/060-Deployment-strimzi-cluster-operator.yaml:81]: I guess some of these things were not supposed to be committed?

Comment (ppatierno) [reply packaging/install/cluster-operator/060-Deployment-strimzi-cluster-operator.yaml]: You were too fast commenting, I was just reverting back changes :-)

**Comment by ppatierno** [comment]
> Should it be a Draft until the proposal is approved to make sure we don't merge it by mistake?

Ok, I will move it to Draft also adding the do-not-merge label :-)

**Comment by ppatierno** [comment]
Moving this PR out of draft because the proposal https://github.com/strimzi/proposals/blob/main/057-run-zk-kraft-clusters-parallel.md  is now merged.

**Review by fvaleri — APPROVED** [review]

**Review by scholzj — APPROVED** [review]

**Comment by ppatierno** [comment]
@tombentley can you have a look at this please?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaReconciler.java
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaReconciler.java:200]: This expression same expression is being used in the file above. Perhaps we can make encapsulate it in method of ReconcilerUtils, i.e. add `ReconcilerUtils.kraftEnabled(featureGates, kafkaCr)`?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorWithPoolsTest.java
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorWithPoolsTest.java:1367]: Call the variable `kraftEnabledKafka` ?

**Review by tombentley — APPROVED** [review]
Thanks, I left a couple of minor comments, but LGTM.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaReconciler.java
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaReconciler.java:200]: Yeah, I think it makes totally sense.

Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaReconciler.java:200]: Actually this is the same we have for node pools. Not having a dedicated method helps to see immediately you are checking both feature gate being enabled and annotation being set, which could be hidden behind just one method. It's not a big deal, I would leave as it is.

**Pull request merged**
